src/main/java/com/aws/greengrass/deployment/KernelUpdateDeploymentTask.java

```diff
@@ -38,49 +42,72 @@ public class KernelUpdateDeploymentTask implements DeploymentTask {
     private final Logger logger;
     private final Deployment deployment;
     private final ComponentManager componentManager;
+    private final ExecutorService executorService;
+    private final CompletableFuture<DeploymentResult> deploymentResultCompletableFuture;
 
     /**
      * Constructor for DefaultDeploymentTask.
      *
-     * @param kernel Kernel instance
-     * @param logger Logger instance
-     * @param deployment Deployment instance
+     * @param kernel           Kernel instance
+     * @param logger           Logger instance
+     * @param deployment       Deployment instance
      * @param componentManager ComponentManager instance
+     * @param executorService  ExecutorService instance
      */
     public KernelUpdateDeploymentTask(Kernel kernel, Logger logger, Deployment deployment,
-                                      ComponentManager componentManager) {
+                                      ComponentManager componentManager, ExecutorService executorService) {
         this.kernel = kernel;
         this.deployment = deployment;
         this.logger = logger.dfltKv(DEPLOYMENT_ID_LOG_KEY, deployment.getGreengrassDeploymentId());
         this.componentManager = componentManager;
+        this.executorService = executorService;
+        this.deploymentResultCompletableFuture = new CompletableFuture<>();
     }
 
     @SuppressWarnings({"PMD.AvoidDuplicateLiterals"})
     @Override
     public DeploymentResult call() {
+        executorService.execute(this::waitForServicesToStart);
+        DeploymentResult result;
+        try {
+            result = deploymentResultCompletableFuture.get();
+        } catch (InterruptedException | ExecutionException | CancellationException e) {
+            return null;
+        }
+        componentManager.cleanupStaleVersions();
+        return result;
+
+    }
+
+    private void waitForServicesToStart() {
         Deployment.DeploymentStage stage = deployment.getDeploymentStage();
         try {
             List<GreengrassService> servicesToTrack =
                     kernel.orderedDependencies().stream().filter(GreengrassService::shouldAutoStart)
                             .filter(o -> !kernel.getMain().equals(o)).collect(Collectors.toList());
             long mergeTimestamp = kernel.getConfig().lookup("system", "rootpath").getModtime();
-            logger.atDebug().kv("serviceToTrack", servicesToTrack).kv("mergeTime", mergeTimestamp)
+
+            logger.atInfo().kv("serviceToTrack", servicesToTrack).kv("mergeTime", mergeTimestamp)
                     .log("Nucleus update workflow waiting for services to complete update");
-            DeploymentConfigMerger.waitForServicesToStart(servicesToTrack, mergeTimestamp, kernel);
+            DeploymentConfigMerger.waitForServicesToStart(servicesToTrack, mergeTimestamp, kernel,
+                    deploymentResultCompletableFuture);
+            if (deploymentResultCompletableFuture.isCancelled()) {
+                return;
```

Comment:
Add a log here that we won't proceed with the deployment as it is cancelled.

Reply:
We already have enough logging.
DeploymentConfigMerger.waitForServicesToStart already logs that deployment was cancelled while waiting for services to start
DeploymentService already logs that we are trying to cancel the deployment then deployment is cancelled

I can add a debug log here for "kernel udpate deployment is cancelled"